### PR TITLE
[TransferEngine] Add topology discovery

### PR DESCRIPTION
**Pull request by wx-csy**
This PR adds automatic topology discovery for CPU/GPU and RDMA devices.

### CODE REVIEW

File: doc/en/transfer-engine.md [hunk 1]
```diff
@@ -61,6 +61,8 @@ For instance, as illustrated in figure above, to transfer data from buffer 0 (as
 To further maximize bandwidth utilization, if a single request's transfer is internally divided into multiple slices if its length exeeds 16KB. 
 Each slice might use a different path, enabling collaborative work among all RDMA NICs.
 
+If you do not want to manually configure the topology matrix, we also provide a function (`mooncake::discoverTopologyMatrix` in `topology.h`) to automatically discover the toplogy between CPU/CUDA and RDMA devices. The automatic discovery mechanism might not always be accurate, and we welcome your feedbacks and improvement ideas!
```
Comment (doujiang24) [inline doc/en/transfer-engine.md]: maybe add a note for TODO could be better? i.e. `TODO: more kinds of devices` after `RDMA devices`?

File: mooncake-transfer-engine/src/topology.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    while ((entry = readdir(dir))) {
+        if (entry->d_name[0] == '.') {
+            continue;
+        }
+
+        std::string device_name = entry->d_name;
+
+        char path[PATH_MAX];
+        char resolved_path[PATH_MAX];
+        snprintf(path, sizeof(path), "/sys/class/infiniband/%s/../..",
```
Comment (alogfans) [inline mooncake-transfer-engine/src/topology.cpp:72]: Is it equivant to `snprintf(path, sizeof(path), "/sys/class/");`?

Comment (wx-csy) [reply mooncake-transfer-engine/src/topology.cpp:72]: Nope, `/sys/class/infiniband/mlx5_X` is a symlink pointing to `/sys/devices/pciXXXX:XX/XXXX:XX:XX.X/infiniband/mlx5_X`, hence the realpath of `/sys/class/infiniband/%s/../..` is the PCI bus id.

Comment (alogfans) [reply mooncake-transfer-engine/src/topology.cpp:72]: Maybe add a comment in code is necessary to avoid incorrect change in the future.

**Comment by alogfans** [comment]
LGTM

**Comment by stmatengss** [comment]
LGTM

**Pull request merged**